<a href="https://colab.research.google.com/github/Ritvik-Bhatia/EDA-HotelBookingAnalysis/blob/main/Hotel_Booking_Analysis_Capstone_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## <b> Have you ever wondered when the best time of year to book a hotel room is? Or the optimal length of stay in order to get the best daily rate? What if you wanted to predict whether or not a hotel was likely to receive a disproportionately high number of special requests? This hotel booking dataset can help you explore those questions!

## <b>This data set contains booking information for a city hotel and a resort hotel, and includes information such as when the booking was made, length of stay, the number of adults, children, and/or babies, and the number of available parking spaces, among other things. All personally identifying information has been removed from the data. </b>

## <b> Explore and analyze the data to discover important factors that govern the bookings. </b>

# 1. Importing necessary libraries to be used in the project


In [ ]:
# libraries to load the dataset and to perfrom analysis on the dataset
import pandas as pd
import numpy as np

# libraries for visualization
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns



# 2. Mounting the drive to access the dataset

## Mounting the drive to access the dataset

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


## Importing Data from Drive

In [ ]:
file_path = '/content/drive/MyDrive/AlmaBetter/capstone_projects/EDA_on_Hotel_Price_Booking/Hotel Bookings.csv'
hoteldata = pd.read_csv(file_path)


# 3. Basic Analisys of the dataset


## checking the first five observations of the dataset by using .head() method.

In [ ]:
hoteldata.head()

hotel  is_canceled  ...  reservation_status  reservation_status_date
0  Resort Hotel            0  ...           Check-Out               2015-07-01
1  Resort Hotel            0  ...           Check-Out               2015-07-01
2  Resort Hotel            0  ...           Check-Out               2015-07-02
3  Resort Hotel            0  ...           Check-Out               2015-07-02
4  Resort Hotel            0  ...           Check-Out               2015-07-03

[5 rows x 32 columns]

## checking the last five observations of the dataset by using .tail() method.

In [ ]:
hoteldata.tail()

hotel  is_canceled  ...  reservation_status  reservation_status_date
119385  City Hotel            0  ...           Check-Out               2017-09-06
119386  City Hotel            0  ...           Check-Out               2017-09-07
119387  City Hotel            0  ...           Check-Out               2017-09-07
119388  City Hotel            0  ...           Check-Out               2017-09-07
119389  City Hotel            0  ...           Check-Out               2017-09-07

[5 rows x 32 columns]

## checking the shape of the dataset by using .shape method.

In [ ]:
hoteldata.shape

(119390, 32)

By using .shape method, we could tell that this Dataset contains a huge number of booking entries with quite a proper amount of details. We have more than 1 lakh booking enteries that we can tell by looking at the number of rows with detailed description that are spread across 32 different features.

## checking the list of indexes to get a basic understanding of Dataset using .columns method. 

In [ ]:
hoteldata.columns

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'adults', 'children', 'babies', 'meal',
       'country', 'market_segment', 'distribution_channel',
       'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'reserved_room_type',
       'assigned_room_type', 'booking_changes', 'deposit_type', 'agent',
       'company', 'days_in_waiting_list', 'customer_type', 'adr',
       'required_car_parking_spaces', 'total_of_special_requests',
       'reservation_status', 'reservation_status_date'],
      dtype='object')

## checking some basic information of data in Dataset and details about Data types in different columns of the dataset using .info() method.


In [ ]:
hoteldata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

## checking the missing-values in the dataset and displaying the percentage of data missing in the columns.


In [ ]:
hoteldata.isnull().sum()[hoteldata.isnull().sum()!=0]/len(hoteldata)*100

children     0.003350
country      0.408744
agent       13.686238
company     94.306893
dtype: float64

Observations after looking at the datatypes and null values:


There are one lakh nineteen thousand three hundred and ninety rows in the dataset which means there are 119390 entries to 32 different indexes.

The columns of the dataset are of different datatypes mostly "object", "integer" or "float" type. Also we can see and observe that a few columns in the dataset have missing values. We need to deal with them by either not using those indexes or removing them from the database.

The float type columns are - "children", "company", "agent" and "adr" but according to my observation this data must have been of Int datatype as number of children cannot be a float so looking at the data we found that there are some null values which makes this data - Float as float datatype supports the storage of NaN values. 

Out of these four columns :

  1. children - contains information about number of children in the dataset.
  
  2. company - contains information about company's ID responsible for making the booking or the payment maybe. in this index almost 94% of the values are missing so this data connot provide us any help in making any descision.

  3. agent - might contain the information about the agent code.

  4. adr - Index Name dosen't suggest anything so looking into the values and googling suggests - adr stands for "Average Daily Rate" which is defined by dividing the sum of all lodging transactions by the total number of staying nights.


While dealing with the null-values of these columns, we will also change their datatypes to integer for simplicity.

## Renaming the column adr to average_daily_rate for simplicity


In [ ]:
hoteldata.rename(columns={'adr':'average_daily_rate'}, inplace = True)

In [ ]:
# checking if the change is successful.
hoteldata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

## Let's drop the column company which has around 95% of the observations as missing and printing the dataset.

In [ ]:
hoteldata.drop('company', axis=1, inplace=True)
hoteldata

hotel  is_canceled  ...  reservation_status  reservation_status_date
0       Resort Hotel            0  ...           Check-Out               2015-07-01
1       Resort Hotel            0  ...           Check-Out               2015-07-01
2       Resort Hotel            0  ...           Check-Out               2015-07-02
3       Resort Hotel            0  ...           Check-Out               2015-07-02
4       Resort Hotel            0  ...           Check-Out               2015-07-03
...              ...          ...  ...                 ...                      ...
119385    City Hotel            0  ...           Check-Out               2017-09-06
119386    City Hotel            0  ...           Check-Out               2017-09-07
119387    City Hotel            0  ...           Check-Out               2017-09-07
119388    City Hotel            0  ...           Check-Out               2017-09-07
119389    City Hotel            0  ...           Check-Out               2017-09-07

[119390 rows x 31 columns]